### ISSUE TITLE
Animate Slide Imageviews

### ISSUE THREAD

**Issue by Xstar97** [PaoloRotolo/AppIntro#223]
You given us the option to ask for permissions which is awesome!
however if the user is sliding across the screen with their fingers & do not use any of the navigation buttons to go to the next slide...the permission dialog wont pop up, but if I press the button to go the next slide until I am leaving the desired slide...it pops ups...have you encountered this?

**Comment by avluis** [comment]
This is a behavior I observed as well when I implemented the library in the app so I'll let @Andrew-Quebe chime in on this;
There have been plenty of merges and fixes towards different issues with the library so this may be part of that, not 100% certain.

**Comment by Xstar97** [comment]
I am using appintro2 since the UI looks better

but what will happen if I use this?
will it disable swiping? or...?
pager.setPagingEnabled(false);

also could you add my tool in your showcase?(I am a new dev...so trying to advertise where I can)

**Comment by avluis** [comment]
`pager.setPagingEnabled(false);` will disable swipe, so only use it as a last resort.

As for your application, do you have the Google Play listing for it?

**Comment by Xstar97** [comment]
I will think about it as only about 10% of my users have android 6 anyway...also I didnt want to post my link unless I had permission...so thanks!
https://play.google.com/store/apps/details?id=com.mcs.txi

**Comment by avluis** [comment]
Just want to note - you can disable pagination just for the permission slide itself - then re-enable it (while making sure that your users cannot scroll back, or simply tracking if the request has already been made) if you wish to keep pagination on.

**Comment by Andrew-Quebe** [comment]
The `askForPermissions` method was intended to be used when your app talks about a certain permission it needs. The process is: explain to the user about said permission > when the next button is pressed, show system permission dialog > regardless of the users grant or denial of permission, advance to the next slide.

Swiping is **automatically disabled** because of the way swipe listeners work...if you don't disable swipe, the permission screen will pop up the instant the user swipes the slide, even the slightest bit.

This is something that many people have asked about because they want swiping to be a thing even if they ask for permissions in the intro. Until swipe listeners perform better, or someone who's better at code than I am comes up with a solution, swiping with your fingers must be disabled, and users must press the navigation buttons.

With regard to your other issue about the dialog popping up _after_ the slide has been advanced, I haven't experienced this issue, and thought that I coded it specifically _not_ to do just that. You should press the next button, and the dialog should pop up. Then, when the system returns that the user has closed the permission dialog, the library advances the slide.

Perhaps show us your full Activity code and we'll see if there are any problems.

**Comment by Xstar97** [comment]
No.. No... The problem is that the permission dialog WONT pop up if the
user swipes... Only after pressing the button to navigate to the slide
makes it work...
For example if I ask for permissions on slide 3...it won't be asked until I
leave that slide which is fine & I don't mind one bit...but could you at
least try to add a piece of code that's like

"Until the next slide is ready to be displayed..ask for the permissions
after swiping fully to next slide"

I am new to developing & that's probably the best explanation you are going
to get out of me...

**Comment by Xstar97** [comment]
Also do you automatically disable swiping if too many permissions are
asked?... Because my earlier version of my tool at once I was using 6
permissions & asking 2 on each slide...& couldn't swipe with my mouse
(Android TV) when testing my tool

**Comment by Andrew-Quebe** [comment]
I'm not sure how you have swiping enabled if you're asking for permissions...that shouldn't be a thing. I'll look into this and talk with @PaoloRotolo to see if we can fix it.

**Comment by Xstar97** [comment]
I launched this from the thread in the code example

My intro!
also I added my permission to slide change & on next pressed to see if that will work(allow the user to swipe through the slides & ask for permissions)
I also moved the shared prefs boolean here because I need my users to go through all the slide & not just exit the app on first launched & disabling the intro when launch again....
also I was wondering...I tried to animate the image view on one slide with mutliples images & just couldn't work...does your library not allow this?
-I don't need to animate the imageview anymore, but some clarification could be nice

import android.*;
import android.content.*;
import android.os.*;
import android.preference.*;
import com.github.paolorotolo.appintro.*;
import com.mcs.txi.*;
import com.mcs.txi.appintro.mainintro.*;

import com.mcs.txi.R;

public class Intro extends AppIntro2
{
    @Override
    public void init(Bundle savedInstanceState) {
		
        addSlide(Intro1.newInstance(R.layout.intro1_main));
        addSlide(Intro2.newInstance(R.layout.intro2_main));
        addSlide(Intro3.newInstance(R.layout.intro3_main));
		addSlide(Intro4.newInstance(R.layout.intro4_main));
		setProgressButtonEnabled(true);
	}
	@Override
    public void onBackPressed()
	{
	}
    @Override
    public void onDonePressed() {
		loadMainActivity();
		SharedPreferences getPrefs = PreferenceManager .getDefaultSharedPreferences(getBaseContext()); 
		SharedPreferences.Editor e = getPrefs.edit(); 
		e.putBoolean("firstStart", false);
		e.apply();
    }

	@Override
    public void onSlideChanged() {
		permissions();
    }

    @Override
    public void onNextPressed() {
		permissions();
    }
    private void loadMainActivity(){
       finish();
	}
	private void permissions(){
		askForPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE}, 3);
	}
}

**Comment by Andrew-Quebe** [comment]
Ah, I see the problem. Rather than calling `permissions()` in the `onNextPressed()` and `onSlideChanged()` methods, just call it in the `init()` method. You also don't really need it in its own method since you're calling it once. 

Below is the corrected code:

```java
import android.*;
import android.content.*;
import android.os.*;
import android.preference.*;
import com.github.paolorotolo.appintro.*;
import com.mcs.txi.*;
import com.mcs.txi.appintro.mainintro.*;

import com.mcs.txi.R;

public class Intro extends AppIntro2
{
    @Override
    public void init(Bundle savedInstanceState) {
		
        addSlide(Intro1.newInstance(R.layout.intro1_main));
        addSlide(Intro2.newInstance(R.layout.intro2_main));
        addSlide(Intro3.newInstance(R.layout.intro3_main));
        
        // Ask for permissions here...
        askForPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE}, 3);

	    addSlide(Intro4.newInstance(R.layout.intro4_main));
	
        // No need for this
        // setProgressButtonEnabled(true);
    }
    @Override
    public void onBackPressed() {
	
    }
    @Override
    public void onDonePressed() {
		loadMainActivity();
		SharedPreferences getPrefs = PreferenceManager .getDefaultSharedPreferences(getBaseContext()); 
		SharedPreferences.Editor e = getPrefs.edit(); 
		e.putBoolean("firstStart", false);
		e.apply();
    }

    @Override
    public void onSlideChanged() {
		// Remove permissions method call
    }

    @Override
    public void onNextPressed() {
		// Remove permissions method call
    }
    private void loadMainActivity() {
       finish();
    }
}
```

What this does is initialize your slides, and, on slide 3, show the permission dialog when the user presses the next button. As said earlier, swiping will be disabled.

**Comment by Xstar97** [comment]
thanks!

**Comment by Andrew-Quebe** [comment]
Seeing that it appears your issue has been solved, I'm going to close this now. If you have any other problems, feel free to open another issue, or reopen this one.

**Comment by Xstar97** [comment]
I just implemented your solution...had to fix a few things that broke...after fixing other things that broke(not caused by your library)
It works as expected now thanks, but I have one question...how can one display an animation of imageviews because I tried multiple ways, but ended up crashing my intro activity...does your library not allow this?

**Comment by avluis** [comment]
What is your approach when animating the ImageView in your layout?
Are you animating via XML, inflating XML manipulating with code?

**Comment by Xstar97** [comment]
I tried both, but couldnt get either to work...I removed the animation code a long time ago
I just want clarification before I implement new projects with appintro..so is it possible?

**Comment by avluis** [comment]
I would need to know what your code looks like since your implementation would dictate what is or not possible.
There is nothing in the library restricting animations that I can find - it's really just a matter of proper implementation.
If I get the time, I'll probably go ahead and try out animating some of the ImageView I have on an app I have AppIntro implemented in, but please don't wait on me for that, give it a go and show us what doesn't work so we can see if there is an issue.

**Comment by Xstar97** [comment]
Okay, I'll look up the tutorials that I used then... Sorry my days & nights are switched so just woke up

**Comment by avluis** [comment]
As it has been a while since the last comment, I will assume the issue fixed or no longer relevant.
If that is not the case, please reopen the issue and comment with appropriate reasoning.
